File: c++/Sorting/BST.cpp
#include <stdio.h>
#include <assert.h>
#include <iostream>
using namespace std;

void print(int data[], int len) {
  for (int i = 0; i < len; i++) {
    cout << data[i] << " ";
  }
  cout << endl;
}

class Node {
  int _val;
  int _sz; //size of the subtree under this node
  int _h; //Height of the subtree under this node
  Node* _parent;
  Node* _left;
  Node* _right;
  void parent(Node *p);
  void left(Node *l);
  void right(Node *r);
  void updateSize();
  void updateHeight();
  Node* root();
  Node* find(int i);

public:
  Node(int val);
  void insert(Node* n);
  void remove(int n);
  void preOrder();
  void inOrder();
  void inOrderRfirst();
  void postOrder();
  void levelOrder();

  int size() { return _sz; };
  int height() { return _h; };
  int min();
  int max();
  int predecessor();
  int successor();
  bool isBST();
};

int Node::predecessor() {
  return 0;
}

int Node::successor() {
  return 0;
}

bool Node::isBST() {
  if (_left != NULL && (_left->_val > _val || !_left->isBST())) {
    return false;
  }
  if (_right != NULL && (_right->_val < _val || !_right->isBST())) {
    return false;
  }
  return true;
}

Node::Node(int v) {
  _val = v;
  _sz = 1;
  _h = 1;
  _parent = NULL;
  _left = NULL;
  _right = NULL;
}

Node* Node::root() {
  Node* n = this;
  while (true) {
    Node* p = n->_parent;
    if (p == NULL) {
      return n;
    }
    n = p;
  }
}

void Node::parent(Node *p) {
  _parent = p;
}

void Node::left(Node *l) {
  _left = l;
  if(l != NULL) {
    l->parent(this);
  }
  updateSize();
  updateHeight();
}

void Node::right(Node *r) {
  _right = r;
  if(r !=NULL) {
    r->parent(this);
  }
  updateSize();
  updateHeight();
}

void Node::updateSize() {
  _sz = 1;
  if (_right != NULL) {
    _sz += _right->size();
  }
  if (_left != NULL) {
    _sz += _left->size();
  }
  if(_parent != NULL) {
    _parent->updateSize();
  }
}

void Node::updateHeight() {
  int lsz = _left == NULL ? -1 : _left->height();
  int rsz = _right == NULL ? -1 : _right->height();
  _h = (lsz > rsz ? lsz : rsz) + 1;
  if(_parent != NULL) {
    _parent->updateHeight();
  }
}

int Node::min() {
  Node* n = this;
  while (true) {
    Node* l = n->_left;
    if (l == NULL) {
      return n->_val;
    }
    n = l;
  }
}

int Node::max() {
  Node* n = this;
  while (true) {
    Node* r = n->_right;
    if (r == NULL) {
      return n->_val;
    }
    n = r;
  }
}

void Node::preOrder() {
  cout << _val << " ";
  if (_left != NULL) {
    _left->preOrder();
  }
  if (_right != NULL) {
    _right->preOrder();
  }
}

void Node::inOrder() {
  //Traverse left side first
  if (_left != NULL) {
    _left->inOrder();
  }
  //Then Traverse self
  cout << _val << " ";
  //Then the right side
  if (_right != NULL) {
    _right->inOrder();
  }
}

void Node::inOrderRfirst() {
  if (_right != NULL) {
    _right->inOrderRfirst();
  }
  cout << _val << " ";
  if (_left != NULL) {
    _left->inOrderRfirst();
  }
}

void Node::postOrder() {
  //Traverse left side first
  if (_left != NULL) {
    _left->postOrder();
  }
  //Then the right side
  if (_right != NULL) {
    _right->postOrder();
  }
  //Then Traverse self
  cout << _val << " ";
}

void Node::levelOrder() {
  //???
}

void Node::insert(Node* n) {
  bool goLeft = n->_val < _val;
  Node *next = goLeft ? _left : _right;
  if (next == NULL) {
    goLeft ? left(n) : right(n);
  } else {
    next->insert(n);
  }
}

Node* Node::find(int n) {
    if (n == _val) {
      return this;
    }
    if (n > _val) {
      return _right->find(n);
    }
    if (n < _val) {
      return _left->find(n);
    }
    return NULL;
}

void Node::remove(int i) {
  Node* n = find(i);
  if (n != NULL) {
    Node* l = n->_left;
    Node* r = n->_right;
    Node* p = n->_parent;
    if (p != NULL) {
      if (p->_left == n) {
          p->left(l != NULL ? l : r);
      } else {
          p->right(l != NULL ? l : r);
      }
    }
    if (l != NULL) {
      l->right(r);
    }
    n->parent(NULL);
    n->left(NULL);
    n->right(NULL);
  }
}

Node* createBST(int data[], int len) {
  Node *root = new Node(data[0]);
  for (int i = 1; i < len; i++) {
    Node* n = new Node(data[i]);
    root->insert(n);
  }
  return root;
}

void sort(int data[], int len, bool asc) {
  Node* bst = createBST(data, 20);
  cout << "PreOrder: ";
  bst->preOrder();
  cout << endl;
  cout << "PostOrder: ";
  bst->postOrder();
  cout << endl;

  cout << "IsBST: " << boolalpha << bst->isBST() << endl;
  cout << "Size: " << bst->size() << endl;
  cout << "Height: " << bst->height() << endl;
  cout << "Min: " << bst->min() << endl;
  cout << "Max: " << bst->max() << endl;
  cout << "InOrder (Sorting): ";
  asc ? bst->inOrder() : bst->inOrderRfirst();
  cout << endl;

  bst->remove(10);
  cout << "After removing node with valye 10: " << endl;
  cout << "IsBST: " << boolalpha << bst->isBST() << endl;
  cout << "Size: " << bst->size() << endl;
  cout << "Height: " << bst->height() << endl;
  cout << "Min: " << bst->min() << endl;
  cout << "Max: " << bst->max() << endl;
  cout << "InOrder (Sorting): ";
  asc ? bst->inOrder() : bst->inOrderRfirst();
  cout << endl;
}

int main(int argc, char const *argv[]) {
  int data[] = {9, 3, 5, 4, 6, 2, 1, 10, 7, 8, 20, 34, 22, 45, 12, 44, 13, 99, 14, 87};
  cout << "Input: ";
  print(data, 20);
  sort(data, 20, false);
  return 0;
}

File: java/BST.java
//Binary trees
public class BST {
  static class Node {
    int data;
    Node left = null;
    Node right = null;
    Node(int data) {
      this.data = data;
    }
  }

  //Find height of the tree
  static int height(Node root) {
    if (root == null) {
      return -1;
    }
    int l = height(root.left);
    int r = height(root.right);
    return l > r ? l + 1 : r + 1;
  }

  private static boolean print(Node root, int level) {
    if (root == null) {
      return false;
    }
    if (level == 0) {
      System.out.print(root.data + " ");
      return true;
    }
    boolean l = print(root.left, level - 1);
    boolean r = print(root.right, level - 1);
    return l || r;
  }

  private static boolean printSpiral(Node root, int level, boolean l2r) {
    if (root == null) {
      return false;
    }
    if (level == 0) {
      System.out.print(root.data + " ");
      return true;
    }
    if (l2r) {
      boolean l = printSpiral(root.left, level - 1, l2r);
      boolean r = printSpiral(root.right, level - 1, l2r);
      return l || r;
    } else {
      boolean r = printSpiral(root.right, level - 1, l2r);
      boolean l = printSpiral(root.left, level - 1, l2r);
      return l || r;
    }
  }

  static void spiralOrder(Node root) {
    int i = 0;
    while (printSpiral(root, i++, i % 2 == 0)) {
      System.out.println();
    }
  }

  //Level order traversal, breadth first
  static void levelOrder(Node root) {
    /*int h = height(root);
    for (int i = 0; i <= h; i++) {
      print(root, i);
      System.out.println("");
    }*/
    int i = 0;
    while (print(root, i++)) {
      System.out.println();
    }
  }

  //Mirror a binary tree
  static void mirror(Node root) {
    if (root == null) {
      return;
    }
    Node t = root.left;
    root.left = root.right;
    root.right = t;
    mirror(root.left);
    mirror(root.right);
  }
  
  public static void main(String[] args) {
    Node root = new Node(4);
    root.left = new Node(2);
    root.right = new Node(6);
    root.left.left = new Node(1);
    root.left.right = new Node(3);
    root.right.left = new Node(5);
    root.right.right = new Node(7);
    System.out.println("Level Order");
    levelOrder(root);
    mirror(root);
    System.out.println("Level Order after mirror");
    levelOrder(root);
    mirror(root);
    System.out.println("Level Order after another mirror");
    levelOrder(root);
    System.out.println("Spiral Order");
    spiralOrder(root); 
  }
}
